Userland/userdel.cpp

```diff
@@ -50,52 +49,94 @@ int main(int argc, char** argv)
     args_parser.add_positional_argument(username, "Login user identity (username)", "login");
     args_parser.parse(argc, argv);
 
-    return delete_user(username, remove_home);
-}
+    char temp_filename[] = "/etc/passwd.XXXXXX";
+    auto fd = mkstemp(temp_filename);
+    if (fd == -1) {
+        perror("failed to create temporary file");
+        return 1;
+    }
 
-int delete_user(const char* username, bool remove_home)
-{
-    const char* temp_filename = "passwd.temp";
-    FILE* temp_file = fopen(temp_filename, "w");
+    FILE* temp_file = fdopen(fd, "w");
     if (!temp_file) {
-        perror("failed to create temporary file");
+        perror("fdopen");
         return 1;
     }
 
-    int error_code = 0;
-    bool user_exists{ false };
+    bool user_exists = false;
+    String home_directory;
+    bool putpwent_error = false;
 
     setpwent();
     for (auto* pw = getpwent(); pw; pw = getpwent()) {
         if (strcmp(pw->pw_name, username)) {
             if (putpwent(pw, temp_file) != 0) {
-                perror("failed to put a entry in the temporay passwd file");
-                error_code = 1;
+                perror("failed to put an entry in the temporary passwd file");
+                putpwent_error = true;
                 break;
             }
         } else {
             user_exists = true;
-            if (remove_home) {
-                if (access(pw->pw_dir, F_OK) != -1 && system(String::format("rm -r %s", pw->pw_dir).characters()) == -1) {
-                    perror("failed to remove the home directory");
-                    error_code = 12;
-                    break;
-                }
-            }
+            if (remove_home)
+                home_directory = pw->pw_dir;
         }
     }
     endpwent();
-    fclose(temp_file);
+
+    if (fclose(temp_file)) {
+        perror("fclose");
+        return 1;
+    }
+
+    if (putpwent_error)
+        return 1;
```

Comment:
Should we not delete the temp file in that case?

Reply:
@bugaevc now deleted.